### ISSUE TITLE
Git diff tab is not restored when reopening vscode

### ISSUE THREAD

**Issue by yj7** [microsoft/vscode#99290]
- VSCode Version: 1.45.1
- OS Version: Win 10 

Steps to Reproduce:

1. Open a project with git configured
2. Open changes of a file using the source control section
3. Close VSCode
4. Relaunch VSCode

All other tabs will be opened except git diff changes tab

Does this issue occur when all extensions are disabled?: Yes/No

**Comment by alexdima** [comment]
Yes, +100 !!

**Comment by joaomoreno** [comment]
@bpasero What will it take for this to happen? This is a simple diff editor with a git document (from a custom file system) on the left side.

I also have a feeling we got this to work recently. 🤔

**Comment by bpasero** [comment]
@joaomoreno you need to register an editor input factory for anything you want to restore.

**Comment by joaomoreno** [comment]
I am (git is) an extension. Should I assign this to you or @alexdima?

**Comment by bpasero** [comment]
Oh wait. Ever since you adopted the file system provider API, the editor input factory is already there for you. However, we do not restore diff editors in general currently. 

Is it safe to assume that a Git URL can simply be opened again in the future?

**Comment by bpasero** [comment]
Ok. Then this essentially is a duplicate of https://github.com/microsoft/vscode/issues/18681

**Comment by joaomoreno** [comment]
But that is out of scope 🤔

**Comment by bpasero** [comment]
Yeah, I am fine making this a feature request and backlog candidate to see if it gets enough votes.

**Comment by bpasero** [comment]
So, with https://github.com/microsoft/vscode/commit/1428d44d172bf6fffddd198c71235df86b6ff0fa pushed, diff editors now have a chance to survive reloads and this works OK unless the diff editor on a git resource is opened:

![recording](https://user-images.githubusercontent.com/900690/84133797-e3d4de80-aa47-11ea-884c-f2e4ef0d9037.gif)

However, if the diff editor is the active one, I am greeted with:

`ERR ENOPRO: No file system provider found for resource 'git:/Users/bpasero/Desktop/test-ts/package.json?%7B%22path%22%3A%22%2FUsers%2Fbpasero%2FDesktop%2Ftest-ts%2Fpackage.json%22%2C%22ref%22%3A%22~%22%7D'`

I think we are trying to restore editors before the extension host is there.

@sandy081 @jrieken @alexdima does this ring a bell? I remember we had nasty deadlocks if the editors would wait to restore before the extension host was ready. Is this simply not supported?

**Comment by jrieken** [comment]
This should work - you can reload successfully with a file from remotes open, e.g with dev containers. This might be something because git activates on `*` which means that we don't wait for it

**Comment by bpasero** [comment]
@jrieken yeah good point, though I thought we might be doing something very special for `vscode-remote` resources so this works.

@joaomoreno can you chime in as well how git gets activated?

**Comment by bpasero** [comment]
I am getting a step closer once I add this to `package.json` for git:

```json
"activationEvents": [
  "*",
  "onFileSystem:git"
]
```

However reloading then I am presented with the following error which makes me think that maybe the git FS provider is maybe not ready yet?

```
 ERR Error: Unable to read file 'git:/Users/bpasero/Desktop/test-ts/CHANGELOG.md?{"path":"/Users/bpasero/Desktop/test-ts/CHANGELOG.md","ref":"~"}' (EntryNotFound (FileSystemError))
    at FileService.doReadAsFileStream (fileService.js:353)
    at processTicksAndRejections (internal/process/task_queues.js:85)
```

**Comment by joaomoreno** [comment]
Hm... good work. I can make the git FS provider wait until the git extension is initialized and has scanned the possible repositories.

**Comment by bpasero** [comment]
**Verification:**
* open a couple of diff editors (from changes view)
* close VSCode
* start again
* verify diff editors restore and show changes properly

**Comment by isidorn** [comment]
Thanks for doing this. Works nicely.
